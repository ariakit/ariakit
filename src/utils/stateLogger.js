/* eslint-disable no-console */
import { diffString } from 'json-diff'

const colorLogger = string => {
  const colorMap = {
    '+': 'green',
    '-': 'red',
  }

  string
    .split('\n')
    .filter(x => !!x)
    .forEach(line => {
      const color = colorMap[line.charAt(0)]
      if (color) {
        console.log(`%c${line}`, `color: ${color}`)
      } else {
        console.log(line)
      }
    })
}

const stateLogger = (prevState, nextState) => {
  const diff = diffString(prevState, nextState, { color: false }).trim()
  if (diff.split('\n').length > 6) {
    console.groupCollapsed('State Update')
  } else {
    console.group('State Update')
  }
  console.log('State', nextState)
  colorLogger(diff)
  console.groupEnd()
}

export default stateLogger
