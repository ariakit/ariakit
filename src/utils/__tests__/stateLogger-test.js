import stateLogger from '../stateLogger'

let logs = []

const fn = v => logs.push(v)

global.console = {
  log: fn,
  group: fn,
  groupEnd: fn,
  groupCollapsed: fn,
}

beforeEach(() => {
  logs = []
})

test('minimal change', () => {
  const left = {
    foo: {
      bar: 'baz',
    },
  }
  const right = {
    foo: {
      bar: 'qux',
    },
  }
  stateLogger(left, right)
  expect(logs.join('\n')).toMatchSnapshot()
})

test('big change', () => {
  const left = {
    foo: {
      bar: 'baz',
      baz: 'qux',
    },
  }
  const right = {
    foo: {
      bar: 'qux',
      baz: 'quux',
    },
  }
  stateLogger(left, right)
  expect(logs.join('\n')).toMatchSnapshot()
})
