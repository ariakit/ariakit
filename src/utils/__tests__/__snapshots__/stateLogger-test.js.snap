// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`big change 1`] = `
"State Update
State
{
   foo: {
%c-    bar: \\"baz\\"
%c+    bar: \\"qux\\"
%c-    baz: \\"qux\\"
%c+    baz: \\"quux\\"
   }
 }
"
`;

exports[`minimal change 1`] = `
"State Update
State
{
   foo: {
%c-    bar: \\"baz\\"
%c+    bar: \\"qux\\"
   }
 }
"
`;
